File: 4_grpc/random_client.cc
#include <iostream>
#include <memory>
#include <string>

#include <cstdlib>

#include <grpc++/grpc++.h>

#include "random.grpc.pb.h"
#include "random.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;

using if3040::grpc::NextRandomRequest;
using if3040::grpc::NextRandomResponse;
using if3040::grpc::RandomService;
using if3040::grpc::SetBoundsRequest;
using if3040::grpc::SetBoundsResponse;

int main(int argc, char *argv[]) {
  if (argc == 1)
    return -1;
  std::string host{argv[1]};

  std::shared_ptr<Channel> channel = grpc::CreateChannel(
      host + ":50051", grpc_impl::InsecureChannelCredentials());
  std::unique_ptr<RandomService::Stub> stub(RandomService::NewStub(channel));

  if (argc == 2) {
    for (int i = 0; i < 10; i++) {
        NextRandomRequest request;
        NextRandomResponse response;
        ClientContext context;
        stub->NextRandom(&context, request, &response);
        printf("%d ", response.random());
	}
    printf("\n");
  }

  if (argc == 4) {
    SetBoundsRequest request;
    SetBoundsResponse response;
    ClientContext context;

    request.set_min(atoi(argv[2]));
    request.set_max(atoi(argv[3]));

    stub->SetBounds(&context, request, &response);
  }

  return 0;
}

File: 4_grpc/random_client.py
import sys
import logging
import grpc

import random_pb2
import random_pb2_grpc


def callNextRandom( hostname ):
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = random_pb2_grpc.RandomServiceStub(channel)
        for i in range(10):
            response = stub.NextRandom(random_pb2.NextRandomRequest())
            print(response.random, end=" ")
        print()

def callSetBounds( hostname, _min, _max ):
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = random_pb2_grpc.RandomServiceStub(channel)
        response = stub.SetBounds(random_pb2.SetBoundsRequest(min=_min, max=_max))


if __name__ == '__main__':
    logging.basicConfig()
    if len( sys.argv ) == 2:
        callNextRandom( sys.argv[1] )
    if len( sys.argv ) == 4:
        callSetBounds( sys.argv[1], int( sys.argv[2] ), int( sys.argv[3] ))

